chore: support ARM architecture

SoManyHs opened the pull request:
Previously, the images pushed were only for amd64 architecture. This
adds support for building arm64 images. Their respective images will now
have a suffix indicating their architecture, e.g.
`amazon/amazon-ecs-local-container-endpoints:latest-arm64`

This change also ensures that go1.15 is being used to build.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

SoManyHs committed:
chore: add support for arm64 arch

Previously, the images pushed were only for amd64 architecture. This
adds support for building arm64 images. Their respective images will now
have a suffix indicating their architecture, e.g.
`amazon/amazon-ecs-local-container-endpoints:latest-arm64`

This change also ensures that go1.15 is being used to build.

SoManyHs committed:
docs: include metadata endpoint v4 in setup-networking docs

SoManyHs committed:
chore: update references to image to include arch suffix

efekarakus commented on Makefile:
```diff
@@ -44,33 +49,60 @@ test:
+		--env ECS_RELEASE=cleanbuild \
+		golang:$(GO_VERSION) make $(AMD_BINARY)
+	docker build -t amazon/amazon-ecs-local-container-endpoints:latest-amd64 .
+	docker tag amazon/amazon-ecs-local-container-endpoints:latest amazon/amazon-ecs-local-container-endpoints:$(TAG)-amd64
+	docker tag amazon/amazon-ecs-local-container-endpoints:latest amazon/amazon-ecs-local-container-endpoints:$(VERSION)-amd64
```
does this need to be like this?
```suggestion
	docker tag amazon/amazon-ecs-local-container-endpoints:latest-amd64 amazon/amazon-ecs-local-container-endpoints:$(TAG)-amd64
	docker tag amazon/amazon-ecs-local-container-endpoints:latest-amd64 amazon/amazon-ecs-local-container-endpoints:$(VERSION)-amd64
```

SoManyHs replied:
ahhh good catch! fixed it both here and for the arm64 images.

efekarakus commented on Makefile:
```diff
@@ -44,33 +49,60 @@ test:
-	docker tag amazon/amazon-ecs-local-container-endpoints:latest amazon/amazon-ecs-local-container-endpoints:$(TAG)
-	docker tag amazon/amazon-ecs-local-container-endpoints:latest amazon/amazon-ecs-local-container-endpoints:$(VERSION)
+		golang:$(GO_VERSION) make $(ARM_BINARY)
+	docker build -t amazon/amazon-ecs-local-container-endpoints:latest-arm64 .
```
If I am following this right, we might need to also update:
https://github.com/awslabs/amazon-ecs-local-container-endpoints/blob/6a6717bf62abf7f173aa528d2ecedda94a3ea7f5/Dockerfile#L8

Right now it's going to use the x86 binary and not arm I think even when running `make release-arm`

SoManyHs replied:
You're right -- I realized that we aren't actually passing in the GOARCH argument to the build_binary.sh, but I think there might be an easier way to do this using docker [buildx](https://docs.docker.com/buildx/working-with-buildx/) so gonna give that a shot.

allisaurus replied:
I think `buildx` is cool but would recommend against relying on it for anything (yet?) as it's still "experimental"

SoManyHs committed:
fix: use correct source image for docker tag

SoManyHs committed:
fix: specify architecture to go compiler

To ensure that the docker images being built are using the binary
compiled for the correct architecture, we are using the `--build-args`
flag to pass in the architecture as a parameter to the Dockerfile.

SoManyHs committed:
chore: add infra code for release pipeline

Use CDK to create a pipeline stack that uses a GitHub webhook and
CodeBuild to publish docker images for amd64 and arm64 architectures to
DockerHub.

TODO: Also publish to ECR Public

SoManyHs commented:
Tested new infrastructure by deploying pipeline to personal [test account](https://github.com/SoManyHs/amazon-ecs-local-container-endpoints/commit/ae58e2e2fbedfce1677345f36ce3a3efb0121814):

<img width="550" alt="Screen Shot 2021-03-24 at 22 07 39" src="https://user-images.githubusercontent.com/29964746/112421982-63891c80-8ced-11eb-89e3-6bb8c1a8455c.png">

Images for both architectures were pushed to personal test DockerHub account, with images built for correct architecture:

<img width="944" alt="Screen Shot 2021-03-25 at 16 19 50" src="https://user-images.githubusercontent.com/29964746/112555735-fecdd080-8d85-11eb-8eed-2c7583729416.png">

SoManyHs committed:
chore: regroup Makefile targets

SoManyHs commented:
Update: successfully tested using `docker manifest` for multiarch image tags:

<img width="870" alt="Screen Shot 2021-03-24 at 23 07 52" src="https://user-images.githubusercontent.com/29964746/112426863-e3b38000-8cf5-11eb-8988-a7df3be735bf.png">

SoManyHs committed:
fix: genericize Make recipes for building images

SoManyHs committed:
update makefile dependency

SoManyHs committed:
add verify stage

PettitWesley commented on Makefile:
```diff
@@ -15,26 +15,73 @@ ROOT := $(shell pwd)
+# environment (e.g. CodeBuild project). Value will either be amd64 or arm64.
+.PHONY: build-image
+build-image:
+	docker run -v $(shell pwd):/usr/src/app/src/github.com/awslabs/amazon-ecs-local-container-endpoints \
+		--workdir=/usr/src/app/src/github.com/awslabs/amazon-ecs-local-container-endpoints \
+		--env GOPATH=/usr/src/app \
+		--env ECS_RELEASE=cleanbuild \
+		golang:$(GO_VERSION) make bin/${ARCH_SUFFIX}/${BINARY_NAME}
```
does GOARCH need to be set here?

SoManyHs replied:
No, I believe we just need to set it in the build_binary script. See this [commit](https://github.com/awslabs/amazon-ecs-local-container-endpoints/pull/59/commits/8ea5381a847c699dedeb135c1a2b3117bbbf3c90#diff-c4914ef202c27bbd0af10a36fb35a813221d15ea12332e61649de07f653966a2R32).

SoManyHs commented:
With successful verification stage added:

Note: -- the `verify` target makes the calls and they are successful, but we may want a more robust way of testing the output of the `curl` commands.

<img width="692" alt="Screen Shot 2021-03-25 at 22 23 56" src="https://user-images.githubusercontent.com/29964746/112586599-533e7380-8db9-11eb-93cb-5dc39cb10170.png">

efekarakus commented on Makefile:
```diff
@@ -44,34 +92,37 @@ test:
-	docker push amazon/amazon-ecs-local-container-endpoints:$(TAG)
-	docker push amazon/amazon-ecs-local-container-endpoints:$(VERSION)
+.PHONY: verify
+verify:
```
woah this is nice!

efekarakus commented on infra/README.md:
```diff
@@ -0,0 +1,31 @@
+
+This package uses the [AWS Cloud Development Kit (AWS)](https://github.com/awslabs/aws-cdk) to model AWS CodePipeline pipelines and to provision them with AWS CloudFormation.
+
+* pipeline.ts: Builds and publishes the base Docker image for the backend API service
```
is this a copy/pasta typo? "backend API service"

SoManyHs replied:
dang caught red handed. will fix

efekarakus commented on infra/pipeline.ts:
```diff
@@ -0,0 +1,153 @@
+      const arch = platform['arch'];
+
+      const buildProject = new codebuild.PipelineProject(this, `BuildImage-${arch}`, {
+        buildSpec: codebuild.BuildSpec.fromSourceFilename('./buildspec.yml'),
```
nit: what do you think of moving the buildspec to the `infra` directory as well?

SoManyHs replied:
So I actually did originally had it in the infra directory, but then moved it to the top level since that's where codebuild typically expects to find it and I thought it'd maybe be easier for it to find the Makefile that way. I also was following what [aws-for-fluent-bit](https://github.com/aws/aws-for-fluent-bit) was doing with where they were keeping their buildspec files, but maybe I was making incorrect assumptions about the codebuild file system. Do you think there's other advantages to putting it under the infra folder?

efekarakus replied:
nah it's fine to keep them at the root, i was just asking to keep the buildspec closer to the codebuild project declaration

efekarakus commented on infra/pipeline.ts:
```diff
@@ -0,0 +1,153 @@
+        }
+      });
+
+      buildProject.addToRolePolicy(new iam.PolicyStatement({
```
Do we need ECR permissions? I might be misunderstanding it but it looks like we are only pushing and pulling from Dockerhub so I wonder if we can delete those permissions 🤔

SoManyHs replied:
We don't currently, but I kept them for when we eventually include ECR repos into our pipeline, which was a suggestion I received from @samuelkarp.

SoManyHs replied:
See: 20f2349b480a30872863370a450be978ae987c8c

efekarakus commented on infra/pipeline.ts:
```diff
@@ -0,0 +1,153 @@
+          "ecr:UploadLayerPart",
+          "ecr:CompleteLayerUpload",
+          "ecr:PutImage",
+          "secretsmanager:GetSecretValue",
```
Do you know if it's possible to only limit this action to the `com.amazonaws.ec2.madison.dockerhub.amazon-ecs-local-container-endpoints.credentials` resource?

SoManyHs replied:
See: 20f2349b480a30872863370a450be978ae987c8c

efekarakus commented on infra/pipeline.ts:
```diff
@@ -0,0 +1,153 @@
+    }
+  }
+
+  // TODO Build stage for creating manifest
```
do we care about this? maybe we can just delete this todo

SoManyHs replied:
It was mentioned as a fast-follow action -- aws-for-fluent-bit includes the manifest creation as part of their [buildspec](https://github.com/aws/aws-for-fluent-bit/blob/mainline/buildspec.yml#L28-L32), and it's mentioned also as a recommended practice in this AWS devops [blog post](https://aws.amazon.com/blogs/devops/creating-multi-architecture-docker-images-to-support-graviton2-using-aws-codebuild-and-aws-codepipeline/).

SoManyHs committed:
docs: update infra readme

SoManyHs committed:
chore: scope down codebuild project permissions
